impala/_thrift_api.py

```diff
@@ -346,23 +381,23 @@ def get_http_transport(host, port, http_path, timeout=None, use_ssl=False,
         except AttributeError:
             auth = base64.encodestring(user_password).decode().strip('\n')
 
-        transport.setCustomHeaders({'Authorization': 'Basic %s' % auth})
+        transport.setGetCustomHeadersFunc(get_custom_headers)
 
     elif auth_mechanism == 'GSSAPI':
         # For GSSAPI over http we need to dynamically generate custom request headers.
-        def get_auth_headers(auth_cookie):
+        def get_custom_headers(cookie_headers):
             import kerberos
-            if auth_cookie:
-                cookie_value = auth_cookie.output(attrs=['value'], header='' ).strip()
-                return {'Cookie': cookie_value}
+            if cookie_headers:
```

Comment:
Here we have the same issue. If one of the cookies is valid but the auth cookie has expired, the authentication will fail.

Perhaps, you should do the same and add the "Authorization" header to the request regardless of the "Cookie" header.

Note that the GSSAPI logic was implemented this way originally because generating the "Authorization" header value can be a bit slow and I wanted to improve performance.

Please, reconsider the logic here. Maybe there's a better approach that is both fast and safe?

Reply:
Same solution as LDAP.  Add "Authorization" header if has_auth_cookie is False.